CareKitUI/CareKitUI/iOS/Task/ChecklistTaskView.swift

```diff
@@ -0,0 +1,307 @@
+                    // If this is the last view in the VStack, add padding to the bottom.
+                    .if(instructions == nil && isCardEnabled && isContentPadded) { $0.padding([.bottom]) }
+                
+                instructions?
+                    .font(.caption)
+                    .foregroundColor(.secondary)
+                    .if(isCardEnabled) { $0.padding([.horizontal, .bottom]) }
+            }
+        }
+    }
+
+    // MARK: - Init
+
+    /// Create an instance.
+    /// - Parameters:
+    ///   - instructions: Instructions text to display under the header.
+    ///   - header: Header to inject at the top of the card. Specified content will be stacked vertically.
+    ///   - content: View to inject under the header. Specified content will be stacked vertically.
+    public init(
+        instructions: Text? = nil,
+        @ViewBuilder header: () -> Header,
+        @ViewBuilder content: () -> Content
+    ) {
+        self.init(
+            isHeaderPadded: false,
+            isContentPadded: false,
+            instructions: instructions,
+            header: header,
+            content: content
+        )
+    }
+
+    init(
+        isHeaderPadded: Bool,
+        isContentPadded: Bool,
+        instructions: Text? = nil,
+        @ViewBuilder header: () -> Header,
+        @ViewBuilder content: () -> Content
+    ) {
+        self.isHeaderPadded = isHeaderPadded
+        self.isContentPadded = isContentPadded
+        self.instructions = instructions
+        self.header = header()
+        self.content = content()
+    }
+}
+
+public extension ChecklistTaskView where Header == _ChecklistTaskViewHeader {
+    /// Create an instance.
+    /// - Parameters:
+    ///   - title: Title text to display in the header.
+    ///   - detail: Detail text to display in the header.
+    ///   - instructions: Instructions text to display under the header.
+    ///   - content: View to inject under the header. Specified content will be stacked vertically.
+    init(
+        title: Text,
+        detail: Text? = nil,
+        instructions: Text? = nil,
+        @ViewBuilder content: () -> Content
+    ) {
+        self.init(
+            isHeaderPadded: true,
+            isContentPadded: false,
+            instructions: instructions,
+            header: {
+                _ChecklistTaskViewHeader(title: title, detail: detail)
+            },
+            content: content
+        )
+    }
+}
+
+public extension ChecklistTaskView {
+    /// Create an instance.
+    /// - Parameters:
+    ///   - instructions: Instructions text to display under the header.
+    ///   - items: Items to display in the checklist.
+    ///   - action: Action to perform when a checklist item is tapped.
+    ///   - header: Header to inject at the top of the card. Specified content will be stacked vertically.
+    init<Item : Completable, ChecklistContent : View>(
```

Comment:
The ability to customize the checklist content is great! 

Currently the view is generic over the `Content`, which allows a developer to replace the entire checklist. When creating the view with custom `Content`, the initializer and layout will look very similar to some other views we have in framework already (see `InstructionsTaskView` when providing a custom `Header` and `Footer`). 

I think it's safe to tie this view to the checklist style, and only allow for customization of rows in the checklist. What do you think? The structure can look something like this:

```swift
public struct ChecklistTaskView<Header: View, ItemContent: View> {

    init<Item: Identifiable>(
        instructions: Text? = nil,
        items: [Item],
        @ViewBuilder header: () -> Header,
        @ViewBuilder itemContent: (Item) -> ItemContent
    ) {
        ...
    }
}

public struct _ChecklistTaskViewItemContent: View {
    ...
}

public struct ChecklistTaskViewItem: Identifiable {
    var id: String
    var title: String
    var isComplete: Bool
}

public extension ChecklistTaskView where Header == HeaderView {

    init<Item: Identifiable>(
        title: Text,
        detail: Text? = nil,
        instructions: Text? = nil,
        items: [Item],
        @ViewBuilder itemContent: (Item) -> ItemContent
    ) {
        ...
    }
}

public extension ChecklistTaskView where ItemContent == _ChecklistTaskViewItemContent {

    init(
        instructions: Text? = nil,
        items: [ChecklistTaskViewItem],
        @ViewBuilder header: () -> Header,
    ) {
        ...
    }
}

public extension ChecklistTaskView where Header == HeaderView, ItemContent == _ChecklistTaskViewItemContent {

    init(
        title: Text,
        detail: Text? = nil,
        instructions: Text? = nil,
        items: [ChecklistTaskViewItem],
    ) {
        ...
    }
}
```

Reply:
Hey, @gavirawson-apple. Thank you, those are great suggestions. You mentioned something I honestly found a bit odd with the current architecture. IMO, the current SwiftUI views are _too_ generic with _too_ specific names. I think it would be good to have some generic views and then wrap them with the more specific views. For example, the following code:

```swift
InstructionsTaskView(header: { Text("Header") }, footer: { Text("Footer") })
SimpleTaskView(header: { Text("Header") }, detailDisclosure: { Text("Detail Disclosure") })
NumericProgressTaskView(header: { Text("Header") }, content: { Text("Content") })
LabeledValueTaskView(header: { Text("Header") }, detailDisclosure: { Text("Detail Disclosure") })
```

Produces the following views:

<img width="295" alt="Capture d’écran 2020-07-21 à 08 29 05" src="https://user-images.githubusercontent.com/5126193/88049882-3ec02200-cb2c-11ea-856e-5550b7ef05b8.png">

As you can see `InstructionsTaskView` and `NumericProgressTaskView` behave identically and its view contents don't have any connection with their type names. Same thing for `SimpleTaskView` and `LabeledValueTaskView`. I think the best path would be to create some rudimentary types, maybe something like `HCardView` and `VCardView`, then build the more specific types with more specific initializers on top of these rudimentary types?

Reply:
I also see version 2.1 hasn't been released yet. I really think this should be addressed before we release 2.1.

Reply:
If you, @gavirawson-apple and @erik-apple, think this makes sense. I can create another PR around this.

Reply:
Thinking more about it, I believe we don't even need something like `HCardView` or `VCardView`, since it would be super trivial to replicate its behavior. I think we should just make the task views behave more specifically. Exposing `CardView` with `HeaderView` already provides most of the flexibility users need.

Reply:
That's a really good point that we have thought about on our end as well. It definitely seems odd. Ultimately, I think the answer lies in your and the community's hands. Let's take the `InstructionsTaskView` as an example. Does exposing just the `HeaderView` and `CardView` allow for enough customization points?

I can offer our thought process that ultimately led to the decision of creating more generic containers like the `Footer` in the `InstructionsTaskView`:

- When designing the API for a view, we try to think about the most likely spots a developer will want to customize. There's a very wide range of use cases to which CareKit is being applied, so we try to lean toward a customizable and open API.

- Say a developer wants to customize the completion button in the `InstructionsTaskView`. Maybe their branding dictates a certain button style that is different from the one provided. If we do not expose the customizable `Footer`, they will have to recreate the view in CareKitUI themself. If they would like a synchronized view as well, they would have to create an associated view in CareKit. That can lead to a lot of extra code and burden on the developer for something as simple as swapping out the button style.

- Currently the `InstructionsTaskView` and the `NumericProgressTaskView` look identical in their most generic form.  However, it's possible that one day in the future the `NumericProgressTaskView ` may take on new UI elements that differentiates it from the `InstructionsTaskView `. And those UI elements might not be customizable through the public API.  If that happens, it will probably be a bit clearer why the `Content` in the `NumericProgressTaskView` is customizable. For example, maybe the view takes on some progress bars in the footer that demonstrate some weekly and monthly statistics for a user's progress towards a goal, for more of a bird's eye view of progress. With the API architected as it is today, we can make that change to the view without breaking the API. 

Those are a few reasons that led us to the architecture we have today. But I'm interested in hearing your thoughts! All feedback is super appreciated and helps us learn how developers are consuming the framework, ultimately leading to a better API!